feat: реализация HTTP сервера для API менеджера задач

Alex-Galvinauskas opened the pull request:
- Добавлен HttpTaskServer с REST API эндпоинтами для задач, эпиков, подзадач
- Реализованы обработчики CRUD операций: TasksHandler, EpicsHandler, SubtasksHandler
- Добавлены сервисные эндпоинты: HistoryHandler, PrioritizedHandler
- Создан базовый обработчик BaseHttpHandler с общей HTTP функциональностью
- Добавлены адаптеры Gson для сериализации LocalDateTime и Duration
- Настроены маршруты сервера и обработка ошибок
- Реализован graceful shutdown с runtime hooks

Эндпоинты:
- GET/POST/DELETE /tasks, /tasks/{id}
- GET/POST/DELETE /epics, /epics/{id}, /epics/{id}/subtasks
- GET/POST/DELETE /subtasks, /subtasks/{id}
- GET /history - история просмотров задач
- GET /prioritized - приоритетный список задач

Alex-Galvinauskas committed:
feat: реализация HTTP сервера для API менеджера задач

- Добавлен HttpTaskServer с REST API эндпоинтами для задач, эпиков, подзадач
- Реализованы обработчики CRUD операций: TasksHandler, EpicsHandler, SubtasksHandler
- Добавлены сервисные эндпоинты: HistoryHandler, PrioritizedHandler
- Создан базовый обработчик BaseHttpHandler с общей HTTP функциональностью
- Добавлены адаптеры Gson для сериализации LocalDateTime и Duration
- Настроены маршруты сервера и обработка ошибок
- Реализован graceful shutdown с runtime hooks

Эндпоинты:
- GET/POST/DELETE /tasks, /tasks/{id}
- GET/POST/DELETE /epics, /epics/{id}, /epics/{id}/subtasks
- GET/POST/DELETE /subtasks, /subtasks/{id}
- GET /history - история просмотров задач
- GET /prioritized - приоритетный список задач

Alex-Galvinauskas committed:
fix: добавлены зависимости Mockito и исправлен CI

Alex-Galvinauskas committed:
fix: Добавить корректные зависимости Mockito в pom.xml

Alex-Galvinauskas committed:
Merge branch 'sprint_9-solution-http-api' of https://github.com/Alex-Galvinauskas/java-kanban into sprint_9-solution-http-api

Alex-Galvinauskas committed:
fix: api-tests.yml

Alex-Galvinauskas committed:
fix: pom.xml

Alex-Galvinauskas committed:
test: заменить mockito на assertJ

Alex-Galvinauskas committed:
test: исправить на чистый Junit

maks-avilow commented on src/taskmanager/app/server/handlers/TasksHandler.java:
```diff
@@ -0,0 +1,160 @@
+                int id = Integer.parseInt(idParam);
+                Optional<Task> task = taskManager.getTaskById(id);
+
+                if (task.isPresent()) {
```
Отсутствует обработка исключения NotFoundException через выброс исключения. 
ТЗ:
«Код можно сделать ещё лаконичнее, пробрасывая NotFoundException в TaskManager. Тогда в обработчиках не нужно проверять экземпляр Task на null — можно обрабатывать сразу исключение. Реализуйте такой подход.» 

Ошибка в коде:
В обработчиках (TasksHandler, EpicsHandler, SubtasksHandler) вручную проверяется:
if (task.isPresent()) {
    sendSuccess(...);
} else {
    sendNotFound(...);
}
При этом в TaskManager (и его реализациях) никаких исключений не выбрасывается при отсутствии сущности — возвращается Optional.empty().  Получается требование ТЗ не выполнено: ты не реализовал проброс исключений из TaskManager, а используешь явную проверку Optional. 
Последствия: 
-код дублируется в каждом обработчике,
-нарушается рекомендация ТЗ по упрощению логики через исключения
В методах getTaskById, getEpicById, getSubTaskById в TaskManager выбрасывать NotFoundException, если сущность не найдена
В обработчиках не проверять Optional, а оборачивать вызов в try-catch

maks-avilow commented on src/taskmanager/app/server/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,198 @@
+                return;
+            }
+
+            List<SubTask> subtasks = taskManager.getEpicSubtasks(epicId);
```
Но в интерфейсе TaskManager метод getEpicSubtasks(int epicId) возвращает List.of() (пустой список):
@Override
public List<SubTask> getEpicSubtasks(int epicId) {
    return List.of();
}
При этом рабочий метод — getSubTasksByEpicId(int epicId), который реализован корректно в InMemoryTaskManager. 
Результат: эндпоинт /epics/1/subtasks всегда возвращает пустой список, даже если подзадачи есть
Заменить в EpicsHandler:
List<SubTask> subtasks = taskManager.getSubTasksByEpicId(epicId);
И удалить или реализовать метод getEpicSubtasks в TaskManager (лучше удалить, если он дублирует)

maks-avilow reviewed (changes requested):
Александр привет! давай немного доработаем до полностью верного варианта

Alex-Galvinauskas committed:
refactor: внести изменения для соответствия ТЗ

 - добавить класс NotFoundException
 - добавить try-catch в обработчики
 - в EpicsHandler заменить getEpicSubtasks на getSubTasksByEpicId(epicId);
- удалить в TaskManager дублирующий  метод getEpicSubtasks

Alex-Galvinauskas committed:
refactor: оптимизировать импорты

maks-avilow reviewed (approved):
зачет
